### PR TITLE
Add asset pack manager and asset packs

### PR DESCRIPTION
**Pull request by IntelOrca**
Adds the asset pack manager and asset packs with initial support for audio asset packs.

**Screenshots:**
image
image

**Video demo:**
https://cdn.discordapp.com/attachments/694586451500859469/987101753247928411/assetpacks_audio.mp4

**Example asset packs:**
[assetpacks.zip](https://github.com/OpenRCT2/OpenRCT2/files/9594424/assetpacks.zip)
Unzip them into `[docs]/openrct2/assetpack`

They are basically the ones from the CI of the open music and open sound repos, but last builds have expired.

### CODE REVIEW

**Comment by karst** [comment]
The game currently closes immediately upon starting up on Windows 10 64 bit

**Comment by IntelOrca** [comment]
@karst I can't reproduce. It does crash when I load your saved game, but I think that might just be a corrupted save. The dump shows that it was doing volume control and ogg decoding at the time. Can you provide me more details as to your setup and whether it crashed on the title screen?

**Comment by karst** [comment]
The crash disappears when you create the assetpack folder manually. It currently won't create it and if it doesn't exist the game will close before even getting to the title screen.

I think the game currently doesn't create an assetpack folder if you don't have one.

**Comment by IntelOrca** [comment]
@karst thanks, that should be simple to fix

**Comment by IntelOrca** [comment]
@karst I have hopefully fixed the issue

**Comment by karst** [comment]
that fixed it!

**Comment by karst** [comment]
I have found one more issue which has to do with reloading, it's a small one. If you have a continuous looping sound, like rain, the weather first has to turn to sunny again before you can actually hear the rain sound when changing the asset pack.

File: src/openrct2/AssetPackManager.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    auto context = GetContext();
+    auto env = context->GetPlatformEnvironment();
+    auto assetPackDirectory = fs::u8path(env->GetDirectoryPath(DIRBASE::USER, DIRID::ASSET_PACK));
+
+    // Create directory if it doesn't exist
+    std::error_code ec;
+    if (!fs::is_directory(assetPackDirectory, ec))
+    {
+        fs::create_directory(assetPackDirectory, ec);
+    }
```
Comment (Broxzier) [inline src/openrct2/AssetPackManager.cpp:76]: We have helper function `Platform::EnsureDirectoryExists` for this.

File: src/openrct2/scripting/bindings/game/ScConfiguration.hpp [hunk 2]
```diff
@@ -199,13 +199,8 @@ namespace OpenRCT2::Scripting
...
                 {
                     auto& localisationService = GetContext()->GetLocalisationService();
-                    auto language = localisationService.GetCurrentLanguage();
-                    auto locale = "";
-                    if (language >= 0 && static_cast<size_t>(language) < std::size(LanguagesDescriptors))
-                    {
-                        locale = LanguagesDescriptors[language].locale;
-                    }
-                    duk_push_string(ctx, locale);
+                    auto locale = localisationService.GetCurrentLanguageLocale();
```
Comment (Gymnasiast) [inline src/openrct2/scripting/bindings/game/ScConfiguration.hpp:202]: This change does not appear to be related to this PR, could you clarify?

Comment (IntelOrca) [reply src/openrct2/scripting/bindings/game/ScConfiguration.hpp:202]: The method I created to get the current locale is used by asset pack window. The other place this should be used is this plugin code.

**Comment by Gymnasiast** [comment]
From watching the video, it seems a bit confusing that there are asset packs for RCT2 graphics, but not for sound or music. That makes it a bit unclear that they override certain sounds/music from the base pack.

**Comment by karst** [comment]
May I reference to Minecraft and their resourcepack system?
image

In Minecraft the default textures are always on. You cannot turn them off (turning them off is grey'd out). The list is laid out in the way of priority (The ones above have priority of below). That means everything that replaces the same thing will be overriden by what's on top. 
I think it'd be a good visual to have the base graphics, soundeffects and music loaded by default. Until OpenSound and OpenGraphics are done this will be RollerCoaster Tycoon 2 + New additions (like the new coaster types).

**Comment by Gymnasiast** [comment]
That’s wasn’t really my point though - it was the inconsistency of having a default pack for graphics, but not sound or music.

**Review by Gymnasiast — COMMENTED** [review]
Ok, I tested this. It seems to work nicely for sound effects and music. :+1: 

I would propose a few changes to the asset pack window:
* [x] Make the ticks into checkboxes, like the Tile Inspector, rather than the flag (more straightforward and also quicker)
* [x] Create a "reload" glyph
* [x] Add tooltips

I haven’t specifically looked into the issues that @karst mentioned.

**Comment by IntelOrca** [comment]
> From watching the video, it seems a bit confusing that there are asset packs for RCT2 graphics, but not for sound or music. That makes it a bit unclear that they override certain sounds/music from the base pack.

> That’s wasn’t really my point though - it was the inconsistency of having a default pack for graphics, but not sound or music.

Sorry I don't really follow, can you elaborate or provide examples of how it should be?

**Comment by karst** [comment]
I don't know how hard it is to implement, (and if it has impact on performance), but it'd be nice if the Asset Pack in use would be stored in the cache so you can change something in the zip file and just press reload to hear the changes (without getting an error that the file is currently in use).

**Review by Gymnasiast — APPROVED** [review]

**Comment by AaronVanGeffen** [comment]
Since the look of the window has been updated, could we get another screenshot here, for posterity?

**Comment by Gymnasiast** [comment]
afbeelding

**Comment by karst** [comment]
> State of asset pack doesn't save after closing the game, you need to reload the assetpacks before they become active again (defaults to base game audio again).

For some odd reason this only occurs on my laptop, I cannot reproduce this on my desktop. Is it possible this might have to do with the base game install being RCTC instead of RCT2TT? That's the only difference I can think of with my laptop. Either way it's not the biggest of deals currently and I am happy with how everything is implemented right now.

**Comment by Gymnasiast** [comment]
@karst That’s certainly possible. I think it’s to investigate that a bit further and create an issue for it.

**Comment by Broxzier** [comment]
Can't respond to image files, so putting it here: the reload icon has a stray pixel on the bottom-left side of the circle.

image

**Comment by Broxzier** [comment]
Putting the photoshop file for the reload icon here in case anyone wants to make more changes. Note that this is not the final version, as @oli414 made some shading changes without this file:

[reload-icon.psd.zip](https://github.com/OpenRCT2/OpenRCT2/files/9736214/reload-icon.zip)
